### PR TITLE
CLOUDP-70011: mongocli iam user(s) invite [flags]

### PR DESCRIPTION
**Pull request by andreaangiolillo**
## Proposed changes

_Jira ticket:_ CLOUDP-70011

## Checklist

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run `make fmt` and formatted my code

## Further comments

Atlas
```
./bin/mongocli iam users invite --country IE --username xnaymer.cawaale4@packetpan.com --password 4ndr34221993. --orgRole 5e429e7706822c6eac4d5970:ORG_MEMBER --email xnaymer.cawaale4@packetpan.com --firstName test --lastName test
The user 'xnaymer.cawaale4@packetpan.com' has been invited.
Invited users do not have access to the project until they accept the invitation.
```
Screenshot 2020-08-31 at 14 24 01

OM
```
./bin/mongocli iam users invite --username xnaymer.cawaale4@packetpan.com --password 4ndr34221993. --orgRole 5f4cef34c864506b4bc5fe76:ORG_MEMBER --email xnaymer.cawaale4@packetpan.com --firstName test --lastName tes -P ops
The user 'xnaymer.cawaale4@packetpan.com' has been invited.
Invited users do not have access to the project until they accept the invitation.
```

Screenshot 2020-08-31 at 14 46 02

### CODE REVIEW

File: internal/cli/iam/users/invite_test.go [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+		Username: "testUser",
+	}
+	opts := &InviteOpts{
+		store:       mockStore,
+		country:     "",
+		mobile:      "",
+		email:       "",
+		lastName:    "",
+		firstName:   "",
+		password:    "",
```
Comment (gssbzn) [inline internal/cli/iam/users/invite_test.go:43]: [nit] you can ommit this, not setting them or setting them to empty string is the same

File: internal/usage/usage.go [hunk 2]
```diff
@@ -31,8 +31,10 @@ const (
 	Status                          = "Alert's status."
 	Until                           = "Acknowledged until a date."
 	Limit                           = "Number of items per page."
-	Username                        = "Username for authenticating to MongoDB."
```
Comment (gssbzn) [inline internal/usage/usage.go:34]: [nit] this is for database user, we may need a new description for app users

**Review by gssbzn — CHANGES_REQUESTED** [review]
This looks great, I left a couple of small nits but nothing too major

File: internal/cli/iam/users/invite.go [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+			i++
+		}
+
+		return roles, nil
+	}
+
+	return nil, nil
+}
+
+func newUserOrgRole(role []string) *opsmngr.UserRole {
```
Comment (gssbzn) [inline internal/cli/iam/users/invite.go:158]: [nit] sorry if this was not clear from my previous comment but the idea of passing role, was to do the splitting logic here, `role` as an array of string has no meaning here, either make it a single string and do the split here or makit it two separate `var`s,  `id` and `role`

File: internal/cli/iam/users/invite.go [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+			}
+			roles[i] = newUserOrgRole(value)
+			i++
+		}
+
+		for _, role := range opts.projectRoles {
+			value := strings.Split(role, ":")
+			if len(value) != keyParts {
+				return nil, fmt.Errorf("unexpected role format: %s", role)
+			}
```
Comment (gssbzn) [inline internal/cli/iam/users/invite.go:147]: it's easier to see now that this can be a function given it's used 4 times

**Review by gssbzn — CHANGES_REQUESTED** [review]
two small follow ups

File: internal/cli/iam/users/invite.go [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+	store        store.UserCreator
+}
+
+func (opts *InviteOpts) init() error {
+	var err error
+	opts.store, err = store.New(config.Default())
+	return err
+}
+
+func (opts *InviteOpts) createUserView() (*store.UserView, error) {
```
Comment (gssbzn) [inline internal/cli/iam/users/invite.go:53]: [nit]
```suggestion
func (opts *InviteOpts) newUserView() (*store.UserView, error) {
```

File: internal/store/users.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+//go:generate mockgen -destination=../mocks/mock_users.go -package=mocks github.com/mongodb/mongocli/internal/store UserCreator
+
+type UserCreator interface {
+	CreateUser(*UserView) (interface{}, error)
+}
+
+type UserView struct {
+	opsmngr.User
+	AtlasRoles   []atlas.AtlasRole `json:"roles"`
+	MobileNumber string            `json:"mobileNumber"`
```
Comment (gssbzn) [inline internal/store/users.go:35]: I think the ops manager API accepts `mobileNumber` so we may need to add this to the go client,

File: internal/store/users.go [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+	"go.mongodb.org/ops-manager/opsmngr"
+)
+
+//go:generate mockgen -destination=../mocks/mock_users.go -package=mocks github.com/mongodb/mongocli/internal/store UserCreator
+
+type UserCreator interface {
+	CreateUser(*UserView) (interface{}, error)
+}
+
+type UserView struct {
```
Comment (gssbzn) [inline internal/store/users.go:32]: [nit]
```suggestion
type UserRequest struct {
```

File: internal/store/users.go [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+	case config.OpsManagerService, config.CloudManagerService:
+		user := &opsmngr.User{
+			Username:     user.User.Username,
+			Password:     user.User.Password,
+			FirstName:    user.User.FirstName,
+			LastName:     user.User.LastName,
+			EmailAddress: user.User.EmailAddress,
+			Roles:        user.User.Roles,
+		}
+		result, _, err := s.client.(*opsmngr.Client).Users.Create(context.Background(), user)
```
Comment (gssbzn) [inline internal/store/users.go:64]: [q] wouldn't you be able to do 
```suggestion
		result, _, err := s.client.(*opsmngr.Client).Users.Create(context.Background(), user.User)
```

File: internal/store/users.go [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+	AtlasRoles   []atlas.AtlasRole `json:"roles"`
+	MobileNumber string            `json:"mobileNumber"`
+	Country      string            `json:"country"`
+}
+
+// CreateUser encapsulates the logic to manage different cloud providers
+func (s *Store) CreateUser(user *UserView) (interface{}, error) {
+	switch s.service {
+	case config.CloudService:
+		user := &atlas.AtlasUser{
```
Comment (gssbzn) [inline internal/store/users.go:43]: [nit] this `user` shadows the `user` from the param

**Review by gssbzn — CHANGES_REQUESTED** [review]
Almost there

File: internal/store/users.go (hunk 6 above)
Comment (andreaangiolillo) [reply internal/store/users.go]: yes, you're right here. I am going to add it.

Comment (andreaangiolillo) [reply internal/store/users.go]: https://github.com/mongodb/go-client-mongodb-ops-manager/pull/44

**Review by gssbzn — APPROVED** [review]
Thanks for bearing with me this long review, great work here, feel free to merge this and do a separate PR updating the client or just ping me for a restamp

**Pull request merged**